File: src/domain/garment-filtering.ts
/**
 * Narrowing and ordering a wardrobe.
 *
 * Pure domain logic, extracted from the list hook so it can be tested without
 * React — and so the native port has one description of the behaviour to follow
 * rather than a re-reading of a component.
 *
 * The database applies the filters it can express (category, availability, a
 * text search across several columns). These are the rest: the ones that need a
 * parsed garment, because they look inside JSON columns or at a value derived
 * from the garment's type.
 */
import type { OccasionOption, SeasonOption } from '../constants/style-filters';
import { getGarmentOccasions } from '../utils/garment-occasions';
import { getGarmentColorPalette } from '../utils/garment-fields';
import type { Garment } from '../types';

export type GarmentSortOption = 'newest' | 'oldest';

export interface GarmentFilter {
  subcategory?: string;
  season?: SeasonOption;
  occasion?: OccasionOption;
  brand?: string;
  size?: string;
  color?: string;
}

/** Case-insensitive "contains", with a blank needle matching everything. */
function contains(haystack: string | null | undefined, needle: string): boolean {
  return (haystack ?? '').toLowerCase().includes(needle.toLowerCase().trim());
}

export function filterGarments(garments: Garment[], filter: GarmentFilter = {}): Garment[] {
  return garments.filter(garment => {
    if (filter.subcategory) {
      const subs = garment.subcategories.length > 0
        ? garment.subcategories
        : (garment.subcategory ? [garment.subcategory] : []);
      if (!subs.includes(filter.subcategory)) return false;
    }

    if (filter.season) {
      const tags = garment.tags.map(tag => tag.toLowerCase());
      if (!tags.includes(filter.season)) return false;
    }

    // Occasion is derived from the garment's type, not stored as a tag.
    if (filter.occasion && !getGarmentOccasions(garment).includes(filter.occasion)) return false;

    if (filter.brand && !contains(garment.brand, filter.brand)) return false;
    if (filter.size && !contains(garment.size, filter.size)) return false;

    if (filter.color && !getGarmentColorPalette(garment).includes(filter.color)) return false;

    return true;
  });
}

/**
 * Compare two timestamps, tolerating their absence.
 *
 * `created_at` is declared non-null but really can be null: an install upgraded
 * through the ALTER path gets the column without NOT NULL, since SQLite cannot
 * add a NOT NULL column without a default. Dereferencing it threw
 * `Cannot read properties of null` for any wardrobe big enough for the null to
 * land on the right-hand side of a comparison — and the list hook caught that,
 * so the screen showed an *empty wardrobe* instead of an error.
 *
 * An absent timestamp sorts as the earliest possible, so such a garment appears
 * last under 'newest' and first under 'oldest' rather than disappearing.
 */
function compareCreatedAt(a: Garment, b: Garment): number {
  return (a.created_at ?? '').localeCompare(b.created_at ?? '');
}

export function sortGarments(
  garments: Garment[],
  sort: GarmentSortOption = 'newest'
): Garment[] {
  const sorted = [...garments];
  return sort === 'oldest'
    ? sorted.sort(compareCreatedAt)
    : sorted.sort((a, b) => compareCreatedAt(b, a));
}

File: native/presentation/src/main/kotlin/com/wardrobapp/presentation/GarmentFiltering.kt
package com.wardrobapp.presentation

import com.wardrobapp.data.GarmentRecord
import com.wardrobapp.domain.Occasion
import com.wardrobapp.domain.Season
import com.wardrobapp.domain.occasions

/**
 * Narrowing and ordering a wardrobe.
 *
 * The database applies the filters it can express -- category, availability, a
 * text search across several columns. These are the rest: the ones that need a
 * parsed garment, because they look inside what were JSON columns or at a value
 * derived from the garment's type.
 */

enum class GarmentSort { NEWEST, OLDEST }

data class GarmentFilter(
    val subcategory: String? = null,
    val season: Season? = null,
    val occasion: Occasion? = null,
    val brand: String? = null,
    val size: String? = null,
    val color: String? = null,
)

/** Case-insensitive "contains", with a blank needle matching everything. */
private fun contains(haystack: String?, needle: String): Boolean =
    (haystack ?: "").lowercase().contains(needle.lowercase().trim())

fun List<GarmentRecord>.filterBy(filter: GarmentFilter): List<GarmentRecord> = filter {
    if (filter.subcategory != null && !it.effectiveSubcategories.contains(filter.subcategory)) {
        return@filter false
    }

    if (filter.season != null && !it.tags.map(String::lowercase).contains(filter.season.tag)) {
        return@filter false
    }

    // Occasion is derived from the garment's type, not stored as a tag.
    if (filter.occasion != null && !it.toDomain().occasions().contains(filter.occasion)) {
        return@filter false
    }

    if (filter.brand != null && !contains(it.brand, filter.brand)) return@filter false
    if (filter.size != null && !contains(it.size, filter.size)) return@filter false

    if (filter.color != null && !it.palette.contains(filter.color)) return@filter false

    true
}

/**
 * Order by when a garment was added.
 *
 * `createdAt` is nullable because an install upgraded through the ALTER path
 * really can have no timestamp -- SQLite cannot add a NOT NULL column without a
 * default. The TypeScript declared it non-null and dereferenced it, which threw
 * for any wardrobe big enough for the null to reach the right-hand side of a
 * comparison; the list hook swallowed that, so the screen showed an *empty
 * wardrobe*. An absent timestamp sorts as the earliest possible, so such a
 * garment appears last under NEWEST and first under OLDEST rather than
 * disappearing.
 */
fun List<GarmentRecord>.sortedBy(sort: GarmentSort = GarmentSort.NEWEST): List<GarmentRecord> {
    val byCreatedAt = compareBy<GarmentRecord> { it.createdAt ?: "" }
    return when (sort) {
        GarmentSort.OLDEST -> sortedWith(byCreatedAt)
        GarmentSort.NEWEST -> sortedWith(byCreatedAt.reversed())
    }
}
